### PR TITLE
Lookup-usage fixes

### PR DESCRIPTION
**Pull request by findepi**
This PR consists of:
* code simplification by replacing `Optional<Lookup>` with either normal `Lookup` or `noLookup()` stub
* a change to `RuleTester` so that `doesNotFire()` tests are carried using same setup/env as positive tests

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeSearcher.java [hunk 1]
```diff
@@ -13,7 +13,6 @@
  */
 package com.facebook.presto.sql.planner.optimizations;
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PlanNodeSearcher.java]: you need to sort commits

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/RuleAssert.java [hunk 2]
```diff
@@ -91,33 +92,29 @@ public RuleAssert on(Function<PlanBuilder, PlanNode> planProvider)
 
     public void doesNotFire()
     {
-        SymbolAllocator symbolAllocator = new SymbolAllocator(symbols);
-        Optional<PlanNode> result = inTransaction(session -> rule.apply(plan, x -> x, idAllocator, symbolAllocator, session));
+        RuleApplication ruleApplication = applyRule();
 
-        if (result.isPresent()) {
+        if (ruleApplication.result.isPresent()) {
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/RuleAssert.java]: maybe replace `result.isPresent()` with something at higher abstract level, like `wasRuleApplied()`?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/RuleAssert.java [hunk 3]
```diff
@@ -91,33 +92,29 @@ public RuleAssert on(Function<PlanBuilder, PlanNode> planProvider)
...
             fail(String.format(
                     "Expected %s to not fire for:\n%s",
                     rule.getClass().getName(),
-                    inTransaction(session -> PlanPrinter.textLogicalPlan(plan, symbolAllocator.getTypes(), metadata, costCalculator, session, 2))));
+                    inTransaction(session -> PlanPrinter.textLogicalPlan(plan, ruleApplication.types, metadata, costCalculator, session, 2))));
```
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/RuleAssert.java]: you can also extract this long line as method so you would not change it in the other place

**Comment by findepi** [comment]
> I would squash all the commits related to noLookup

They are almost independent. I hope it's easier to review that way.

**Comment by findepi** [comment]
Squashed & rebased. @martint, this is ready for review.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/RuleAssert.java [hunk 4]
```diff
@@ -138,11 +135,30 @@ public void matches(PlanMatchPattern pattern)
...
     }
 
+    private RuleApplication applyRule()
+    {
+        SymbolAllocator symbolAllocator = new SymbolAllocator(symbols);
+        Memo memo = new Memo(idAllocator, plan);
+        Lookup lookup = Lookup.from(memo::resolve);
+        Optional<PlanNode> result = inTransaction(session -> rule.apply(memo.getNode(memo.getRootGroup()), lookup, idAllocator, symbolAllocator, session));
+
+        return new RuleApplication(
```
Comment (martint) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/RuleAssert.java]: These arguments can all go on the same line

File: presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/RuleAssert.java [hunk 5]
```diff
@@ -153,4 +169,30 @@ public void matches(PlanMatchPattern pattern)
                     return transactionSessionConsumer.apply(session);
                 });
     }
+
+    private static class RuleApplication
+    {
+        private final Memo memo;
```
Comment (martint) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/rule/test/RuleAssert.java]: This doesn't seem to be used anywhere

**Comment by findepi** [comment]
@martint , thanks for review! comments applied

**Comment by martint** [comment]
Merged, thanks!

**Pull request closed without merging**